File: sorts/quick_sort/quick-sort.rb
def quick_sort! arr, left = 0, right = arr.length - 1, &comp
  # left is greater than right when there are two
  # elems remaining and left starts at "partition + 1"
  return if left >= right 

  pivot_idx = (left + right) / 2
  # move pivot to left-most index so comparisons
  # to remaining elements are sequential
  swap! arr, pivot_idx, left

  num_smaller = 0
  pivot = arr[left]
  (left + 1..right).each do |i|
    current_el = arr[i]
    current_el_precedes = (block_given? ? comp.call(current_el, pivot)
                                        : current_el < pivot)
    if current_el_precedes
      num_smaller += 1
      swap! arr, i, left + num_smaller
    end
  end
  
  # right partition separates pivot's smaller
  # els from pivot's greater els
  right_partition = left + num_smaller
  swap! arr, left, right_partition
  quick_sort! arr, left, right_partition - 1, &comp
  quick_sort! arr, right_partition + 1, right, &comp
end

def swap! arr, i, j
  arr[i], arr[j] = arr[j], arr[i]
end

File: sorts/quick_sort/c/quick-sort.c
#include "quick-sort.h"
#include <stdarg.h>

int ascending(void* aPtr, void* bPtr) {
  int a = *(int*) aPtr, b = *(int*) bPtr;
  return a > b;
}

int descending(void* aPtr, void* bPtr) {
  int a = *(int*) aPtr,
      b = *(int*) bPtr;
  return a < b;
}

void quickSortStep(
      void* array[], const int left, const int right, comp_func);

void _swap(void* arr[], const int i, const int j);

void quickSort(void* array[], const int len, int comp_given, ...) {
  va_list argP;

  va_start(argP, comp_given);

  if (comp_given) {
    comp_func comparor = va_arg(argP, comp_func);
    quickSortStep(array, 0, len - 1, comparor);
  } else {
    quickSortStep(array, 0, len - 1, (int (*)(void*, void*)) ascending);
  }

  va_end(argP);
}

void quickSortStep(void* array[], const int left, const int right,
      comp_func comp
    ) {
  int pivot_idx, i, right_partition;

  // left is greater than right when there are two
  // els remaining and left starts at "partition + 1"
  if (left >= right) {
    return;
  }

  pivot_idx = (left + right) / 2;
  // move pivot to left-most index so comparisons
  // to remaining elements are sequential
  _swap(array, left, pivot_idx);
  // right partition separates pivot's smaller
  // els from pivot's greater els
  right_partition = left;

  for (i = left + 1; i <= right; i++) {
    if (comp(array[left], array[i])) {
      _swap(array, ++right_partition, i);
    }
  }

  _swap(array, left, right_partition);

  quickSortStep(
      array, left, right_partition - 1,
      (int (*)(void*, void*)) comp
    );
  quickSortStep(
      array, right_partition + 1, right,
      (int (*)(void*, void*)) comp
    );
}

void _swap(void* arr[], const int i, const int j) {
  void* temp = arr[i];
  arr[i] = arr[j];
  arr[j] = temp;
}

